File: backend/utils/alloy_utils.py
import json

def get_graph_data(alloy_instance: json):
  nodes = set()
  edges = []

  def get_atoms_from_sig(d):
    def search_nested_dict(d):
      if isinstance(d, dict):
        for k, v in d.items():
          if k == 'atom':
            if isinstance(v, list):
              for item in v:
                if 'label' in item:
                  nodes.add(str(item['label']).replace('$', ''))
            elif 'label' in v:
              nodes.add(str(v['label']).replace('$', ''))
          elif isinstance(v, dict) or isinstance(v, list):
            search_nested_dict(v)
      elif isinstance(d, list):
        for i in d:
          search_nested_dict(i)
    search_nested_dict(d)

  if 'sig' in alloy_instance["alloy"]["instance"]:
    get_atoms_from_sig(alloy_instance["alloy"]["instance"]["sig"])
  
  def process_field(field):
    relation = field.get('label')
    tuple_field = field.get('tuple', {})
    if isinstance(tuple_field, dict):
        tuples = [tuple_field]
    else:
        tuples = tuple_field
    for t in tuples:
        if isinstance(t, dict):
            atoms = t.get('atom', [])
            if len(atoms) >= 2:
                source_label = atoms[0].get('label')
                target_label = atoms[1].get('label')
                print(source_label, target_label)
                if source_label and target_label:
                    nodes.add(source_label.replace('$', ''))
                    nodes.add(target_label.replace('$', ''))
                    edges.append(
                        {
                            "data": {
                                "id": f"{source_label}_{target_label}",
                                "label": relation,
                                "source": source_label.replace('$', ''),
                                "target": target_label.replace('$', ''),
                                "relationship": relation,
                            }
                        }
                    )
        elif isinstance(t, list):
            for atom in t:
                if isinstance(atom, dict):
                    source_label = atom.get('label')
                    print(source_label)
                    if source_label:
                        nodes.add(source_label.replace('$', ''))

  if 'field' in alloy_instance["alloy"]["instance"]:
      fields = alloy_instance["alloy"]["instance"]["field"]
      if isinstance(fields, list):
        for field in fields:
          process_field(field)
      else:
        process_field(fields)
  # else:
  #     for instance in alloy_instance["alloy"]["instance"]:
  #       fields = instance["field"]
  #       if isinstance(fields, list):
  #         for field in fields:
  #           process_field(field)
  #       else:
  #         process_field(fields)
  print(nodes)
  print(edges)
  node_list = [{"data": {"id": node, "label": node}} for node in nodes]
  elements = node_list + edges
  specId = alloy_instance["specId"]
  return json.dumps({"elements": elements, "specId": specId})

File: frontend/src/assets/js/alloyUtils.js
export function getGraphData(alloyInstance) {
  const nodes = new Set();
  const edges = [];
  const subsetSigs = new Set();

  function getAtomsFromSig(d) {
    function searchNestedDict(d) {
      if (typeof d === 'object' && d !== null) {
        for (const [k, v] of Object.entries(d)) {
          if (k === 'atom') {
            if (Array.isArray(v)) {
              for (const item of v) {
                if ('label' in item) {
                  nodes.add(String(item['label']).replace('$', ''));
                }
              }
            } else if ('label' in v) {
              nodes.add(String(v['label']).replace('$', ''));
            }
          } else if (typeof v === 'object') {
            searchNestedDict(v);
          }
        }
      } else if (Array.isArray(d)) {
        for (const i of d) {
          searchNestedDict(i);
        }
      }
    }
    searchNestedDict(d);

    // Handle Subset sigs. Issue #9
    const varSigs = d
      .filter(item => item.var === "yes" && item.atom)
      .flatMap(item => {
        const atoms = Array.isArray(item.atom) ? item.atom : [item.atom];
        return atoms.map(atomItem => ({
          atom: atomItem.label.replace('$', ''),  // Remove the $ symbol
          label: item.label,
        }));
      });
    varSigs.forEach(item => {
      subsetSigs.add(item);
    });

  }

  function processField(field) {
    const label = field['label'];
    // relationship is computed from the field label and the types of the atoms
    const atomTypes = [];
    // check length of types array
    if (field['types']['type']) {
      // collect type elements in field['types'] types 
      for (const type of field['types']['type'] || {}) {
        atomTypes.push(type['ID']);
      }
    } else {
      for (const obj of field['types'] || {}) {
        for (const type of obj['type'] || {}) {
          atomTypes.push(type['ID']);
        }
      }
    }
    const relationship = field['label'] + ' (' + atomTypes.join('_') + ')';
    const tupleField = field['tuple'] || {};
    const tuples = Array.isArray(tupleField) ? tupleField : [tupleField];
    for (const t of tuples) {
      if (typeof t === 'object' && t !== null) {
        const atoms = t['atom'] || [];
        if (atoms.length >= 2) {
          if (atoms.length > 2) { // Nested relation according to official Alloy 
            const sourceLabel = atoms[0]['label'].toString();
            const targetLabel = atoms[atoms.length - 1]['label'].toString();
            if (sourceLabel && targetLabel) {
              nodes.add(sourceLabel.replace('$', ''));
              nodes.add(targetLabel.replace('$', ''));
              for (let i = 1; i < atoms.length - 1; i++) {
                const nodeLabel = atoms[i]['label'].toString();
                edges.push({
                  "data": {
                    "id": `${sourceLabel}_${targetLabel}_${label}_[${nodeLabel.replace('$', '')}]`,
                    "label": `${label} [${nodeLabel.replace('$', '')}]`,
                    "source": sourceLabel.replace('$', ''),
                    "target": targetLabel.replace('$', ''),
                    "relationship": relationship,
                  }
                });
              }
            }

          } else {
            const sourceLabel = atoms[0]['label'].toString();
            const targetLabel = atoms[1]['label'].toString();
            if (sourceLabel && targetLabel) {
              nodes.add(sourceLabel.replace('$', ''));
              nodes.add(targetLabel.replace('$', ''));
              edges.push({
                "data": {
                  "id": `${sourceLabel}_${targetLabel}`,
                  "label": label,
                  "source": sourceLabel.replace('$', ''),
                  "target": targetLabel.replace('$', ''),
                  "relationship": relationship,
                }
              });
            }
          }
        } else if (Array.isArray(t)) {
          for (const atom of t) {
            if (typeof atom === 'object' && atom !== null) {
              const sourceLabel = atom['label'].toString();
              if (sourceLabel) {
                nodes.add(sourceLabel.replace('$', ''));
              }
            }
          }
        }
      }
    }
  }

  if ('sig' in alloyInstance) {
    getAtomsFromSig(alloyInstance["sig"]);
  }
  if ('field' in alloyInstance) {
    const fields = Array.isArray(alloyInstance["field"]) ? alloyInstance["field"] : [alloyInstance["field"]];
    for (const field of fields) {
      processField(field);
    }
  }

  const nodeList = Array.from(nodes).map(node => ({ "data": { "id": node, "label": node } }));
  const elements = nodeList.concat(edges);

  // Handle Subset sigs. Issue #9
  const atomLabelMap = new Map();
  subsetSigs.forEach(item => {
    // Group labels by atom
    if (atomLabelMap.has(item.atom)) {
      atomLabelMap.get(item.atom).push(item.label);
    } else {
      atomLabelMap.set(item.atom, [item.label]);
    }
  });
  elements.forEach(element => {
    const id = element.data.id;
    if (atomLabelMap.has(id)) {
      const labels = atomLabelMap.get(id);
      // Update the element's label by concatenating all related labels
      element.data.label = `${id} (${labels.join(', ')})`;
    }
  });
  return elements;
}

export function parseAlloyErrorMessage(error) {
  let message = '';
  if (error.includes('error') && error.includes('.als') && error.includes('line')) {
    message = error.replace(/ in .+\.als/, '');
  } else if (error.includes('The required JNI library cannot be found')) {
    message = 'The required JNI library cannot be found.';
  } else {
    message = error;
  }
  return message;
}

export function getTraceLengthAndBackloop(alloyInstance) {
  let traceLength = null;
  let backloop = null;
  if ('tracelength' in alloyInstance) {
    traceLength = alloyInstance['tracelength'];
  }
  if ('backloop' in alloyInstance) {
    backloop = alloyInstance['backloop'];
  }
  return { traceLength, backloop };
}